### PR TITLE
[5.0]  tinyMCE b/c fixes

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #40622

### Summary of Changes

- @richard67 added a function to transform some J3/4 parameters to their new expected values for tinyMCE v6
- Replaced the same values in the installation SQL files
- Removed the template plugin as it's deprecated
- Introduce a fork of the template plugin, named `jtemplate`
- Patch both the builder for the above change
- Introduce a function `onAjaxTinymce()`, so that templates will be fetched on request rather than on the rendering of the editor (security + perf)
- Removed/renamed plugins and options to satisfy the changes of tinyMCE v6

### Testing Instructions

1. d/l and install a fresh copy of Joomla from the GH PR. Check that tinyMCE works as expected, particularly check the templates functionality.

2. Have a J4 installation and create some more `TinyMCE Sets`. Try to use all the buttons,menus,etc, and also have some html (or txt) files as templates in the `cassiopeia/html/tinymce`. UPGRADE to J5 using the `Custom URL` in the `administrator/index.php?option=com_config&view=component&component=com_joomlaupdate&path=` with a vaule `https://ci.joomla.org/artifacts/joomla/joomla-cms/5.0-dev/40626/downloads/65966/pr_list.xml`. Enure that the configuarations are identical and that the templates are still working.

### Actual result BEFORE applying this Pull Request

Broken Upgrade

### Expected result AFTER applying this Pull Request

No breaks

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

@brianteeman @HLeithner @laoneo

### CODE REVIEW

File: plugins/editors/tinymce/src/PluginTraits/ToolbarPresets.php [hunk 1]
```diff
@@ -64,7 +64,7 @@ public static function getToolbarPreset()
                     'alignleft', 'aligncenter', 'alignright', 'alignjustify', '|',
                     'lineheight', '|',
                     'styleselect', '|',
-                    'formatselect', 'fontfamily', 'fontsizeselect', '|',
+                    'formatselect', 'fontfamily', 'fontsize', '|',
```
Comment (brianteeman) [inline plugins/editors/tinymce/src/PluginTraits/ToolbarPresets.php]: shouldnt formatselect be changed to blocks ? https://www.tiny.cloud/docs/tinymce/6/migration-from-5x/

Comment (dgrammatiko) [reply plugins/editors/tinymce/src/PluginTraits/ToolbarPresets.php]: Yup

File: plugins/editors/tinymce/src/PluginTraits/ToolbarPresets.php [hunk 2]
```diff
@@ -64,7 +64,7 @@ public static function getToolbarPreset()
                     'alignleft', 'aligncenter', 'alignright', 'alignjustify', '|',
                     'lineheight', '|',
                     'styleselect', '|',
```
Comment (brianteeman) [inline plugins/editors/tinymce/src/PluginTraits/ToolbarPresets.php]: should this not be 'styles'

**Comment by brianteeman** [comment]
\plugins\editors\tinymce\src\PluginTraits\DisplayTrait.php

`            $scriptOptions['forced_root_block'] = '';	
`	
> forced_root_block must be a non-empty string and cannot take a value of false.

**Comment by dgrammatiko** [comment]
> forced_root_block must be a non-empty string and cannot take a value of false.

So what should it be?

**Comment by brianteeman** [comment]
no idea

**Comment by brianteeman** [comment]
textpattern options https://www.tiny.cloud/docs/tinymce/6/6.0-release-notes-core-changes/#removed-or-deprecated-options-textpattern

memo to self to review this

**Comment by brianteeman** [comment]
```
            'paste'        => ['label' => 'Paste', 'plugin' => 'paste'],
            'pastetext'    => ['label' => 'Paste as text', 'plugin' => 'paste'],
```

> https://www.tiny.cloud/docs/tinymce/6/6.0-release-notes-core-changes/#changed-plugins-paste

**Comment by brianteeman** [comment]
>             'emoticons'      => ['label' => 'Emoticons', 'plugin' => 'emoticons'],

https://www.tiny.cloud/docs/tinymce/6/6.0-release-notes-core-changes/#new-and-improved-plugins-emoticons

Check but i think that means the label is now "Emoji"

**Comment by dgrammatiko** [comment]
> emoticons

This still exists in the plugin folder

**Comment by brianteeman** [comment]
To Check
[ ] Do we still need any of the language strings PLG_TINY_TOOLBAR_BUTTON_ etc
[ ] Button Seperator not displayed
[ ] Editor has rounded corners
[ ] Emoji/Emoticons
[ ] Text patterns
[ ] Forced root block - think this was something @crystalenka added

File: plugins/editors/tinymce/src/PluginTraits/KnownButtons.php [hunk 3]
```diff
@@ -74,8 +74,7 @@ public static function getKnownButtons()
 
             'cut'          => ['label' => 'Cut'],
             'copy'         => ['label' => 'Copy'],
-            'paste'        => ['label' => 'Paste', 'plugin' => 'paste'],
-            'pastetext'    => ['label' => 'Paste as text', 'plugin' => 'paste'],
```
Comment (brianteeman) [inline plugins/editors/tinymce/src/PluginTraits/KnownButtons.php:78]: why remove this button?

Comment (dgrammatiko) [reply plugins/editors/tinymce/src/PluginTraits/KnownButtons.php:78]: Wrong keystroke

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 4]
```diff
@@ -412,6 +412,14 @@ public function onDisplay(
 
         // Merge the two toolbars for backwards compatibility
         $toolbar = array_merge($toolbar1, $toolbar2);
+        // Rename old toolbar buttons for B/C
```
Comment (HLeithner) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: Can't we write a migration script for this?

**Comment by dgrammatiko** [comment]
> Button Seperator not displayed

That's a visual difference for v6 (there's no separation bar anymore)

**Comment by brianteeman** [comment]
shouldnt jtemplate live in here \build\media_source\plg_editors_tinymce\js\plugins

**Comment by dgrammatiko** [comment]
> shouldnt jtemplate live in here \build\media_source\plg_editors_tinymce\js\plugins

This would require more changes in the builder

**Comment by brianteeman** [comment]
But it would be the correct way forward and then wont require further changes when they actually remove it?

**Comment by dgrammatiko** [comment]
> But it would be the correct way forward and then wont require further changes when they actually remove it?

Ok, the code was moved to the _correct_ folder

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php (hunk 4 above)
Comment (richard67) [reply plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: Done. Dimitris has merged my PR with the migration.

**Comment by richard67** [comment]
@dgrammatiko I've allowed myself to update your branch for this PR to the latest 5.0-dev with the results of the upmerge so that we get a usable update package built by Drone for this PR, and so people can use that for testing the migration. Before that, updating from 4.4 to 5 was broken due to the missing upmerge.

**Comment by dgrammatiko** [comment]
I will update the description, testing instructions, etc, later on or tomorrow

**Comment by brianteeman** [comment]
> > Button Seperator not displayed

> That's a visual difference for v6 (there's no separation bar anymore) 

The _problem_ then is that the visual seperator is still shown in the plugin configuration

File: plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php [hunk 5]
```diff
@@ -304,46 +305,9 @@ public function onDisplay(
...
-                    'url'         => Uri::root(true) . $template_path . '/' . $full_filename,
-                ];
-            }
-        }
+        $jtemplates = !empty($allButtons['jtemplate'])
+            ? Uri::base(true) . '/index.php?option=com_ajax&plugin=tinymce&group=editors&format=json&format=json&template=' . $levelParams->get('content_template_path') . '&' . $csrf . '=1'
+            : false;
 
         // Check for extra plugins, from the setoptions form
         foreach (['wordcount' => 1, 'advlist' => 1, 'autosave' => 1, 'textpattern' => 0] as $pName => $def) {
```
Comment (brianteeman) [inline plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]:
```suggestion
        foreach (['wordcount' => 1, 'advlist' => 1, 'autosave' => 1] as $pName => $def) {
```

textpattern is not an external plugin any more

Comment (dgrammatiko) [reply plugins/editors/tinymce/src/PluginTraits/DisplayTrait.php]: Thanks for researching this

**Comment by brianteeman** [comment]
Can you please update the licence in the tinymce.xml as it is no longer LGPL but is instead MIT

**Comment by brianteeman** [comment]
source code plugin (highlighter) fails with

﻿```

plugin-es5.min.js:1 Uncaught TypeError: Cannot read properties of undefined (reading 'codemirror')
    at Object.t [as onAction] (plugin-es5.min.js:1:153)
    at theme.min.js:4:283552
    at theme.min.js:4:123218
    at theme.min.js:4:283521
    at theme.min.js:4:30451
    at L (theme.min.js:4:2411)
    at run (theme.min.js:4:30436)
    at theme.min.js:4:31303
    at theme.min.js:4:795
    at theme.min.js:4:152163
```

**Comment by dgrammatiko** [comment]
> source code plugin (highlighter) fails with

I've seen it as well, but let's leave this for another PR (it needs also the Codemirror v5 upgrade). I mean patching the plugin for the existing code mirror would be wasted energy as the editor needs an upgrade...

File: installation/sql/mysql/base.sql [hunk 6]
```diff
@@ -282,7 +282,7 @@ INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`,
...
...":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"1":{"access":["6","2"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"0":{"access":["7","4","8"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"1","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""}}},"sets_amount":3,"html_height":"550","html_width":"750"}', '', 3, 0),
```
Comment (richard67) [inline installation/sql/mysql/base.sql]:
```suggestion
(0, 'plg_editors_tinymce', 'plugin', 'tinymce', 'editors', 0, 1, 1, 0, 1, '', '{"configuration":{"toolbars":{"2":{"toolbar1":["bold","underline","strikethrough","|","undo","redo","|","bullist","numlist","|","pastetext"]},"1":{"menu":["edit","insert","view","format","table","tools"],"toolbar1":["bold","italic","underline","strikethrough","|","alignleft","aligncenter","alignright","alignjustify","|","blocks","|","bullist","numlist","|","outdent","indent","|","undo","redo","|","link","unlink","anchor","code","|","hr","table","|","subscript","superscript","|","charmap","pastetext","preview"]},"0":{"menu":["edit","insert","view","format","table","tools"],"toolbar1":["bold","italic","underline","strikethrough","|","alignleft","aligncenter","alignright","alignjustify","|","styles","|","blocks","fontfamily","fontsize","|","searchreplace","|","bullist","numlist","|","outdent","indent","|","undo","redo","|","link","unlink","anchor","image","|","code","|","forecolor","backcolor","|","fullscreen","|","table","|","subscript","superscript","|","charmap","emoticons","media","hr","ltr","rtl","|","cut","copy","paste","pastetext","|","visualchars","visualblocks","nonbreaking","blockquote","jtemplate","|","print","preview","codesample","insertdatetime","removeformat"]}},"setoptions":{"2":{"access":["1"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"1":{"access":["6","2"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"0":{"access":["7","4","8"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"1","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""}}},"sets_amount":3,"html_height":"550","html_width":"750"}', '', 3, 0),
```

Comment (richard67) [reply installation/sql/mysql/base.sql]: "fontselect" => "fontfamily" in toolbar

File: installation/sql/postgresql/base.sql [hunk 7]
```diff
@@ -288,7 +288,7 @@ INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder",
...
...":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"1":{"access":["6","2"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"0":{"access":["7","4","8"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"1","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""}}},"sets_amount":3,"html_height":"550","html_width":"750"}', '', 3, 0),
```
Comment (richard67) [inline installation/sql/postgresql/base.sql]:
```suggestion
(0, 'plg_editors_tinymce', 'plugin', 'tinymce', 'editors', 0, 1, 1, 0, 1, '', '{"configuration":{"toolbars":{"2":{"toolbar1":["bold","underline","strikethrough","|","undo","redo","|","bullist","numlist","|","pastetext"]},"1":{"menu":["edit","insert","view","format","table","tools"],"toolbar1":["bold","italic","underline","strikethrough","|","alignleft","aligncenter","alignright","alignjustify","|","blocks","|","bullist","numlist","|","outdent","indent","|","undo","redo","|","link","unlink","anchor","code","|","hr","table","|","subscript","superscript","|","charmap","pastetext","preview"]},"0":{"menu":["edit","insert","view","format","table","tools"],"toolbar1":["bold","italic","underline","strikethrough","|","alignleft","aligncenter","alignright","alignjustify","|","styles","|","blocks","fontfamily","fontsize","|","searchreplace","|","bullist","numlist","|","outdent","indent","|","undo","redo","|","link","unlink","anchor","image","|","code","|","forecolor","backcolor","|","fullscreen","|","table","|","subscript","superscript","|","charmap","emoticons","media","hr","ltr","rtl","|","cut","copy","paste","pastetext","|","visualchars","visualblocks","nonbreaking","blockquote","jtemplate","|","print","preview","codesample","insertdatetime","removeformat"]}},"setoptions":{"2":{"access":["1"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"1":{"access":["6","2"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"0","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""},"0":{"access":["7","4","8"],"skin":"0","skin_admin":"0","mobile":"0","drag_drop":"1","path":"","entity_encoding":"raw","lang_mode":"1","text_direction":"ltr","content_css":"1","content_css_custom":"","relative_urls":"1","newlines":"0","use_config_textfilters":"0","invalid_elements":"script,applet,iframe","valid_elements":"","extended_elements":"","resizing":"1","resize_horizontal":"1","element_path":"1","wordcount":"1","image_advtab":"1","advlist":"1","autosave":"1","contextmenu":"1","custom_plugin":"","custom_button":""}}},"sets_amount":3,"html_height":"550","html_width":"750"}', '', 3, 0),
```
"fontselect" => "fontfamily" in toolbar

**Comment by HLeithner** [comment]
thanks, can you please create a pr for the issue mentioned by brian?

**Pull request merged**